{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## About\n",
    "This study aims to identify the best method to preprocess and accurately portray mass spectrometric isotopic ratio data from individual amino acids. Experiments with valine (C<sub>5</sub>H<sub>11</sub>NO<sub>2</sub>), proline (C<sub>5</sub>H<sub>9</sub>NO<sub>2</sub>) and hydroxyproline (C<sub>5</sub>H<sub>9</sub>NO<sub>3</sub>) samples has been conducted and four isotopes ratios are of interest (H<sub>1</sub>/H<sub>2</sub>, C<sub>12</sub>/C<sub>13</sub>, N<sub>14</sub>/N<sub>15</sub> and O<sub>16</sub>/O<sub>18</sub>). The goal is find a method to represent the data in a accurate and precise way. Ideally, the ranking of the amino acids should be the same for different isotopic ratios (the accuracy problem) and the variance of the measurement should not be too broad (the precision problem). Specifically, previous attempts has shown that there the ranking of amino acids differ for N and H isotopes, and that carbon isotopic ratios has a high variance (in the order of percentages, when it is expected to be in promille). The carbon measurement problem causes an issue in reproducibility of experiments. These errors could be caused by either measurement and sample preperation errors, or data preprocessing and statistical method. This project focuses solely on the data preprocessing, data processing and statistical methods approach.\n",
    "\n",
    "\n",
    "\n",
    "## Current steps\n",
    "- Precision problem: Find method that reduces the variance of isotopic ratios (specifically interested in finding it for C).\n",
    "- Accuracy problem: Find method that gives stable ranking of amino acids for the different isotopic ratios.\n",
    "- Check coefficient of variance between different replicates, to check normalization.\n",
    "- Check if different transformations could give better results (for example log-transformation.\n",
    "- Find a method to yield meaningful reproducib<sub></sub>le experimental results, without ratios (is there a way to find similar results as above without ratios?)\n",
    "- Use median of ratios, instead of ratios (Check Romans paper).\n",
    "- Remocal of non-statistical outliers: Use subset of medians of data (medians of triplicates or quadruplicates), simulate similar process with gaussian distribution to find the correction factor and apply the correction factor to the romal of non-statistical outliers process. \n",
    "- \n",
    "Goal is to have a methods that produces reproducable experimental results.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## To read:\n",
    "### Papers\n",
    "- [A Researcher's Guide to Mass Spectrometry-based Proteomics](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5198776/)\n",
    "- [Evaluation of Peak-Picking Algorithms for Protein Mass Spectrometry](https://www.researchgate.net/publication/47730346_Evaluation_of_Peak-Picking_Algorithms_for_Protein_Mass_Spectrometry)\n",
    "\n",
    "### Website\n",
    "- [High-resolution accurate mass (HRAM) for metabolomics and lipidomics](https://www.thermofisher.com/se/en/home/industrial/mass-spectrometry/mass-spectrometry-learning-center/mass-spectrometry-applications-area/metabolomics-mass-spectrometry/overview-mass-spectrometry-metabolomics/hram-mass-spectrometry-metabolomics.html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are a lot of softwares for ms preprocessing. What do we actually want to do?\n",
    "\n",
    "Do these software do all three steps (smoothing, baseline correction and peak picking)?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### imports\n",
    "\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/home/ptruong/git/isoMS/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_dir = \"data/Patrick/Experiment1/mzML/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['20200506_DirectInjection_Lumos_Hyp_1.mzML',\n",
       " '20200506_DirectInjection_Lumos_Pro_2.mzML',\n",
       " '20200506_DirectInjection_Lumos_Valine_3.mzML',\n",
       " '20200506_DirectInjection_Lumos_Hyp_2.mzML',\n",
       " '20200506_DirectInjection_Lumos_Valine_2.mzML',\n",
       " '20200506_DirectInjection_Lumos_Hyp_3.mzML',\n",
       " '20200506_DirectInjection_Lumos_Pro_1.mzML',\n",
       " '20200506_DirectInjection_Lumos_Pro_3.mzML',\n",
       " '20200506_DirectInjection_Lumos_Valine_1.mzML']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir(file_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyteomics import mzml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "mzml_file = mzml.read(file_dir + '20200506_DirectInjection_Lumos_Hyp_1.mzML')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'index': 0,\n",
       " 'id': 'controllerType=0 controllerNumber=1 scan=1',\n",
       " 'defaultArrayLength': 1054264,\n",
       " 'scanList': {'count': 1,\n",
       "  'scan': [{'scanWindowList': {'count': 1,\n",
       "     'scanWindow': [{'scan window lower limit': 50.0 m/z,\n",
       "       'scan window upper limit': 200.0 m/z}]},\n",
       "    'scan start time': 0.0018222472 minute,\n",
       "    'filter string': 'FTMS + p NSI Full ms [50.0000-200.0000]',\n",
       "    'ion injection time': 0.282198995352 millisecond}],\n",
       "  'no combination': ''},\n",
       " 'MS1 spectrum': '',\n",
       " 'ms level': 1,\n",
       " 'positive scan': '',\n",
       " 'profile spectrum': '',\n",
       " 'base peak m/z': 132.066116333008 m/z,\n",
       " 'base peak intensity': 1695461760.0 number of detector counts,\n",
       " 'total ion current': 3927511296.0,\n",
       " 'lowest observed m/z': 49.502896162251 m/z,\n",
       " 'highest observed m/z': 202.014287725871 m/z,\n",
       " 'spectrum title': '20200506_DirectInjection_Lumos_Hyp_1.1.1. File:\"20200506_DirectInjection_Lumos_Hyp_1.raw\", NativeID:\"controllerType=0 controllerNumber=1 scan=1\"',\n",
       " 'count': 2,\n",
       " 'm/z array': array([ 49.50289616,  49.50294289,  49.50298961, ..., 202.01351739,\n",
       "        202.01390256, 202.01428773]),\n",
       " 'intensity array': array([0., 0., 0., ..., 0., 0., 0.])}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(mzml_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'index': 1,\n",
       " 'id': 'controllerType=0 controllerNumber=1 scan=3',\n",
       " 'defaultArrayLength': 527694,\n",
       " 'scanList': {'count': 1,\n",
       "  'scan': [{'scanWindowList': {'count': 1,\n",
       "     'scanWindow': [{'scan window lower limit': 50.0 m/z,\n",
       "       'scan window upper limit': 200.0 m/z}]},\n",
       "    'scan start time': 0.186301419733 minute,\n",
       "    'filter string': 'FTMS + p NSI Full ms2 125.0000@hcd50.00 [50.0000-200.0000]',\n",
       "    'ion injection time': 0.255162477493 millisecond,\n",
       "    '[Thermo Trailer Extra]Monoisotopic M/Z:': 0.0}],\n",
       "  'no combination': ''},\n",
       " 'precursorList': {'count': 1,\n",
       "  'precursor': [{'spectrumRef': 'controllerType=0 controllerNumber=1 scan=1',\n",
       "    'isolationWindow': {'isolation window target m/z': 125.0 m/z,\n",
       "     'isolation window lower offset': 75.0 m/z,\n",
       "     'isolation window upper offset': 75.0 m/z},\n",
       "    'selectedIonList': {'count': 1,\n",
       "     'selectedIon': [{'selected ion m/z': 125.0 m/z,\n",
       "       'charge state': 1.0,\n",
       "       'peak intensity': 2830297.25 number of detector counts}]},\n",
       "    'activation': {'beam-type collision-induced dissociation': '',\n",
       "     'collision energy': 50.0 electronvolt}}]},\n",
       " 'MSn spectrum': '',\n",
       " 'ms level': 2,\n",
       " 'positive scan': '',\n",
       " 'profile spectrum': '',\n",
       " 'base peak m/z': 68.049713134766 m/z,\n",
       " 'base peak intensity': 408337952.0 number of detector counts,\n",
       " 'total ion current': 1379620608.0,\n",
       " 'lowest observed m/z': 49.502813063554 m/z,\n",
       " 'highest observed m/z': 202.015486169218 m/z,\n",
       " 'spectrum title': '20200506_DirectInjection_Lumos_Hyp_1.3.3.1 File:\"20200506_DirectInjection_Lumos_Hyp_1.raw\", NativeID:\"controllerType=0 controllerNumber=1 scan=3\"',\n",
       " 'count': 2,\n",
       " 'm/z array': array([ 49.50281306,  49.50290651,  49.50299995, ..., 202.01394548,\n",
       "        202.01471582, 202.01548617]),\n",
       " 'intensity array': array([0., 0., 0., ..., 0., 0., 0.])}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "next(mzml_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2020-11-25 Wednesday\n",
    "### A look on the processed data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.listdir(\"data\")\n",
    "pd.set_option('display.max_rows', 30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_dir = \"data/MS/\"\n",
    "df = pd.read_csv(file_dir + \"20200306_AminoAcid_Pro_1_fit.csv\", sep = \",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqNum</th>\n",
       "      <th>peak</th>\n",
       "      <th>I</th>\n",
       "      <th>mu</th>\n",
       "      <th>sigma</th>\n",
       "      <th>R2</th>\n",
       "      <th>isoshift</th>\n",
       "      <th>theorshift</th>\n",
       "      <th>masserror</th>\n",
       "      <th>isoratio</th>\n",
       "      <th>rt</th>\n",
       "      <th>tic</th>\n",
       "      <th>bpI</th>\n",
       "      <th>bpMZ</th>\n",
       "      <th>ion</th>\n",
       "      <th>n</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>178901.448535</td>\n",
       "      <td>116.070730</td>\n",
       "      <td>0.000269</td>\n",
       "      <td>0.996828</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>P</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>13C</td>\n",
       "      <td>9755.975781</td>\n",
       "      <td>117.074117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.996654</td>\n",
       "      <td>1.003386</td>\n",
       "      <td>1.003355</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>0.054533</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>P</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15N</td>\n",
       "      <td>418.000876</td>\n",
       "      <td>117.067945</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.978855</td>\n",
       "      <td>0.997215</td>\n",
       "      <td>0.997035</td>\n",
       "      <td>0.000180</td>\n",
       "      <td>0.002336</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>P</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>2H</td>\n",
       "      <td>163.643807</td>\n",
       "      <td>117.076463</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.519899</td>\n",
       "      <td>1.005733</td>\n",
       "      <td>1.006277</td>\n",
       "      <td>-0.000544</td>\n",
       "      <td>0.000915</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>P</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>18O</td>\n",
       "      <td>569.345768</td>\n",
       "      <td>118.075025</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.995120</td>\n",
       "      <td>2.004295</td>\n",
       "      <td>2.004246</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.003182</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>P</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>257.237644</td>\n",
       "      <td>132.065977</td>\n",
       "      <td>0.000347</td>\n",
       "      <td>0.997496</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>13C</td>\n",
       "      <td>6.016313</td>\n",
       "      <td>133.069326</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.813113</td>\n",
       "      <td>1.003349</td>\n",
       "      <td>1.003355</td>\n",
       "      <td>-0.000006</td>\n",
       "      <td>0.023388</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>2H</td>\n",
       "      <td>10.525238</td>\n",
       "      <td>133.072418</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.884321</td>\n",
       "      <td>1.006440</td>\n",
       "      <td>1.006277</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.040916</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>18O</td>\n",
       "      <td>2.725352</td>\n",
       "      <td>134.069755</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.752480</td>\n",
       "      <td>2.003778</td>\n",
       "      <td>2.004246</td>\n",
       "      <td>-0.000468</td>\n",
       "      <td>0.010595</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>81.359754</td>\n",
       "      <td>132.102343</td>\n",
       "      <td>0.000343</td>\n",
       "      <td>0.997799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>L</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1</td>\n",
       "      <td>13C</td>\n",
       "      <td>13.107043</td>\n",
       "      <td>133.105277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.976767</td>\n",
       "      <td>1.002935</td>\n",
       "      <td>1.003355</td>\n",
       "      <td>-0.000420</td>\n",
       "      <td>0.161100</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>L</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1</td>\n",
       "      <td>15N</td>\n",
       "      <td>5.149972</td>\n",
       "      <td>133.099589</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.807918</td>\n",
       "      <td>0.997247</td>\n",
       "      <td>0.997035</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.063299</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>L</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>60.416140</td>\n",
       "      <td>175.119584</td>\n",
       "      <td>0.000502</td>\n",
       "      <td>0.967439</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>R</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1</td>\n",
       "      <td>2H</td>\n",
       "      <td>9.268543</td>\n",
       "      <td>176.125581</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.889246</td>\n",
       "      <td>1.005997</td>\n",
       "      <td>1.006277</td>\n",
       "      <td>-0.000280</td>\n",
       "      <td>0.153412</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>R</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>13.821508</td>\n",
       "      <td>147.113317</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.991786</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.742637</td>\n",
       "      <td>1726367104</td>\n",
       "      <td>683092160</td>\n",
       "      <td>116.070732</td>\n",
       "      <td>K</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5916</th>\n",
       "      <td>551</td>\n",
       "      <td>13C</td>\n",
       "      <td>8791.439136</td>\n",
       "      <td>117.074095</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.997454</td>\n",
       "      <td>1.003380</td>\n",
       "      <td>1.003355</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.055839</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>P</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5917</th>\n",
       "      <td>551</td>\n",
       "      <td>15N</td>\n",
       "      <td>347.933041</td>\n",
       "      <td>117.067940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.977363</td>\n",
       "      <td>0.997225</td>\n",
       "      <td>0.997035</td>\n",
       "      <td>0.000190</td>\n",
       "      <td>0.002210</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>P</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5918</th>\n",
       "      <td>551</td>\n",
       "      <td>2H</td>\n",
       "      <td>184.536215</td>\n",
       "      <td>117.076503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.714345</td>\n",
       "      <td>1.005788</td>\n",
       "      <td>1.006277</td>\n",
       "      <td>-0.000488</td>\n",
       "      <td>0.001172</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>P</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5919</th>\n",
       "      <td>551</td>\n",
       "      <td>18O</td>\n",
       "      <td>488.837748</td>\n",
       "      <td>118.075031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.990125</td>\n",
       "      <td>2.004316</td>\n",
       "      <td>2.004246</td>\n",
       "      <td>0.000070</td>\n",
       "      <td>0.003105</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>P</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5920</th>\n",
       "      <td>551</td>\n",
       "      <td>0</td>\n",
       "      <td>45.012392</td>\n",
       "      <td>132.065854</td>\n",
       "      <td>0.000325</td>\n",
       "      <td>0.996065</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5921</th>\n",
       "      <td>551</td>\n",
       "      <td>15N</td>\n",
       "      <td>6.176221</td>\n",
       "      <td>133.063224</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.864769</td>\n",
       "      <td>0.997370</td>\n",
       "      <td>0.997035</td>\n",
       "      <td>0.000335</td>\n",
       "      <td>0.137212</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5922</th>\n",
       "      <td>551</td>\n",
       "      <td>18O</td>\n",
       "      <td>8.718556</td>\n",
       "      <td>134.070006</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.876692</td>\n",
       "      <td>2.004153</td>\n",
       "      <td>2.004246</td>\n",
       "      <td>-0.000093</td>\n",
       "      <td>0.193692</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Hyp</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5923</th>\n",
       "      <td>551</td>\n",
       "      <td>0</td>\n",
       "      <td>26.583910</td>\n",
       "      <td>163.112282</td>\n",
       "      <td>0.000454</td>\n",
       "      <td>0.989100</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Hyl</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5924</th>\n",
       "      <td>551</td>\n",
       "      <td>13C</td>\n",
       "      <td>10.785321</td>\n",
       "      <td>164.115311</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.873716</td>\n",
       "      <td>1.003029</td>\n",
       "      <td>1.003355</td>\n",
       "      <td>-0.000326</td>\n",
       "      <td>0.405709</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Hyl</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5925</th>\n",
       "      <td>551</td>\n",
       "      <td>0</td>\n",
       "      <td>6.754685</td>\n",
       "      <td>166.088562</td>\n",
       "      <td>0.000383</td>\n",
       "      <td>0.982695</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>F</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5926</th>\n",
       "      <td>551</td>\n",
       "      <td>2H</td>\n",
       "      <td>12.584362</td>\n",
       "      <td>167.094526</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.986516</td>\n",
       "      <td>1.005964</td>\n",
       "      <td>1.006277</td>\n",
       "      <td>-0.000313</td>\n",
       "      <td>1.863057</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>F</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5927</th>\n",
       "      <td>551</td>\n",
       "      <td>0</td>\n",
       "      <td>7.391953</td>\n",
       "      <td>147.074913</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.987799</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Q</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5928</th>\n",
       "      <td>551</td>\n",
       "      <td>15N</td>\n",
       "      <td>5.586200</td>\n",
       "      <td>148.072038</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.887294</td>\n",
       "      <td>0.997125</td>\n",
       "      <td>0.997035</td>\n",
       "      <td>0.000090</td>\n",
       "      <td>0.755714</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Q</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5929</th>\n",
       "      <td>551</td>\n",
       "      <td>2H</td>\n",
       "      <td>7.856256</td>\n",
       "      <td>148.081334</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.969122</td>\n",
       "      <td>1.006421</td>\n",
       "      <td>1.006277</td>\n",
       "      <td>0.000144</td>\n",
       "      <td>1.062812</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Q</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5930</th>\n",
       "      <td>551</td>\n",
       "      <td>18O</td>\n",
       "      <td>1.959529</td>\n",
       "      <td>149.079349</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.857968</td>\n",
       "      <td>2.004435</td>\n",
       "      <td>2.004246</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.265089</td>\n",
       "      <td>2998.647960</td>\n",
       "      <td>1490738944</td>\n",
       "      <td>602472128</td>\n",
       "      <td>116.070717</td>\n",
       "      <td>Q</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5931 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seqNum peak              I          mu     sigma        R2  isoshift  \\\n",
       "0          1    0  178901.448535  116.070730  0.000269  0.996828  0.000000   \n",
       "1          1  13C    9755.975781  117.074117       NaN  0.996654  1.003386   \n",
       "2          1  15N     418.000876  117.067945       NaN  0.978855  0.997215   \n",
       "3          1   2H     163.643807  117.076463       NaN  0.519899  1.005733   \n",
       "4          1  18O     569.345768  118.075025       NaN  0.995120  2.004295   \n",
       "5          1    0     257.237644  132.065977  0.000347  0.997496  0.000000   \n",
       "6          1  13C       6.016313  133.069326       NaN  0.813113  1.003349   \n",
       "7          1   2H      10.525238  133.072418       NaN  0.884321  1.006440   \n",
       "8          1  18O       2.725352  134.069755       NaN  0.752480  2.003778   \n",
       "9          1    0      81.359754  132.102343  0.000343  0.997799  0.000000   \n",
       "10         1  13C      13.107043  133.105277       NaN  0.976767  1.002935   \n",
       "11         1  15N       5.149972  133.099589       NaN  0.807918  0.997247   \n",
       "12         1    0      60.416140  175.119584  0.000502  0.967439  0.000000   \n",
       "13         1   2H       9.268543  176.125581       NaN  0.889246  1.005997   \n",
       "14         1    0      13.821508  147.113317  0.000329  0.991786  0.000000   \n",
       "...      ...  ...            ...         ...       ...       ...       ...   \n",
       "5916     551  13C    8791.439136  117.074095       NaN  0.997454  1.003380   \n",
       "5917     551  15N     347.933041  117.067940       NaN  0.977363  0.997225   \n",
       "5918     551   2H     184.536215  117.076503       NaN  0.714345  1.005788   \n",
       "5919     551  18O     488.837748  118.075031       NaN  0.990125  2.004316   \n",
       "5920     551    0      45.012392  132.065854  0.000325  0.996065  0.000000   \n",
       "5921     551  15N       6.176221  133.063224       NaN  0.864769  0.997370   \n",
       "5922     551  18O       8.718556  134.070006       NaN  0.876692  2.004153   \n",
       "5923     551    0      26.583910  163.112282  0.000454  0.989100  0.000000   \n",
       "5924     551  13C      10.785321  164.115311       NaN  0.873716  1.003029   \n",
       "5925     551    0       6.754685  166.088562  0.000383  0.982695  0.000000   \n",
       "5926     551   2H      12.584362  167.094526       NaN  0.986516  1.005964   \n",
       "5927     551    0       7.391953  147.074913  0.000261  0.987799  0.000000   \n",
       "5928     551  15N       5.586200  148.072038       NaN  0.887294  0.997125   \n",
       "5929     551   2H       7.856256  148.081334       NaN  0.969122  1.006421   \n",
       "5930     551  18O       1.959529  149.079349       NaN  0.857968  2.004435   \n",
       "\n",
       "      theorshift  masserror  isoratio           rt         tic        bpI  \\\n",
       "0       0.000000   0.000000  1.000000     0.742637  1726367104  683092160   \n",
       "1       1.003355   0.000032  0.054533     0.742637  1726367104  683092160   \n",
       "2       0.997035   0.000180  0.002336     0.742637  1726367104  683092160   \n",
       "3       1.006277  -0.000544  0.000915     0.742637  1726367104  683092160   \n",
       "4       2.004246   0.000049  0.003182     0.742637  1726367104  683092160   \n",
       "5       0.000000   0.000000  1.000000     0.742637  1726367104  683092160   \n",
       "6       1.003355  -0.000006  0.023388     0.742637  1726367104  683092160   \n",
       "7       1.006277   0.000163  0.040916     0.742637  1726367104  683092160   \n",
       "8       2.004246  -0.000468  0.010595     0.742637  1726367104  683092160   \n",
       "9       0.000000   0.000000  1.000000     0.742637  1726367104  683092160   \n",
       "10      1.003355  -0.000420  0.161100     0.742637  1726367104  683092160   \n",
       "11      0.997035   0.000212  0.063299     0.742637  1726367104  683092160   \n",
       "12      0.000000   0.000000  1.000000     0.742637  1726367104  683092160   \n",
       "13      1.006277  -0.000280  0.153412     0.742637  1726367104  683092160   \n",
       "14      0.000000   0.000000  1.000000     0.742637  1726367104  683092160   \n",
       "...          ...        ...       ...          ...         ...        ...   \n",
       "5916    1.003355   0.000025  0.055839  2998.647960  1490738944  602472128   \n",
       "5917    0.997035   0.000190  0.002210  2998.647960  1490738944  602472128   \n",
       "5918    1.006277  -0.000488  0.001172  2998.647960  1490738944  602472128   \n",
       "5919    2.004246   0.000070  0.003105  2998.647960  1490738944  602472128   \n",
       "5920    0.000000   0.000000  1.000000  2998.647960  1490738944  602472128   \n",
       "5921    0.997035   0.000335  0.137212  2998.647960  1490738944  602472128   \n",
       "5922    2.004246  -0.000093  0.193692  2998.647960  1490738944  602472128   \n",
       "5923    0.000000   0.000000  1.000000  2998.647960  1490738944  602472128   \n",
       "5924    1.003355  -0.000326  0.405709  2998.647960  1490738944  602472128   \n",
       "5925    0.000000   0.000000  1.000000  2998.647960  1490738944  602472128   \n",
       "5926    1.006277  -0.000313  1.863057  2998.647960  1490738944  602472128   \n",
       "5927    0.000000   0.000000  1.000000  2998.647960  1490738944  602472128   \n",
       "5928    0.997035   0.000090  0.755714  2998.647960  1490738944  602472128   \n",
       "5929    1.006277   0.000144  1.062812  2998.647960  1490738944  602472128   \n",
       "5930    2.004246   0.000189  0.265089  2998.647960  1490738944  602472128   \n",
       "\n",
       "            bpMZ  ion     n  \n",
       "0     116.070732    P   NaN  \n",
       "1     116.070732    P   5.0  \n",
       "2     116.070732    P   1.0  \n",
       "3     116.070732    P  10.0  \n",
       "4     116.070732    P   2.0  \n",
       "5     116.070732  Hyp   NaN  \n",
       "6     116.070732  Hyp   5.0  \n",
       "7     116.070732  Hyp  10.0  \n",
       "8     116.070732  Hyp   3.0  \n",
       "9     116.070732    L   NaN  \n",
       "10    116.070732    L   6.0  \n",
       "11    116.070732    L   1.0  \n",
       "12    116.070732    R   NaN  \n",
       "13    116.070732    R  15.0  \n",
       "14    116.070732    K   NaN  \n",
       "...          ...  ...   ...  \n",
       "5916  116.070717    P   5.0  \n",
       "5917  116.070717    P   1.0  \n",
       "5918  116.070717    P  10.0  \n",
       "5919  116.070717    P   2.0  \n",
       "5920  116.070717  Hyp   NaN  \n",
       "5921  116.070717  Hyp   1.0  \n",
       "5922  116.070717  Hyp   3.0  \n",
       "5923  116.070717  Hyl   NaN  \n",
       "5924  116.070717  Hyl   6.0  \n",
       "5925  116.070717    F   NaN  \n",
       "5926  116.070717    F  12.0  \n",
       "5927  116.070717    Q   NaN  \n",
       "5928  116.070717    Q   2.0  \n",
       "5929  116.070717    Q  11.0  \n",
       "5930  116.070717    Q   3.0  \n",
       "\n",
       "[5931 rows x 16 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**seqNum**: Scan Number\n",
    "\n",
    "**peak**: Isotope Element, 0 means it's the monoisotopic peak\n",
    "\n",
    "**I**: Area under the curve of the gaussian fit of the peak (signal intensity)\n",
    "\n",
    "**mu**: Mass unit (m/z)\n",
    "\n",
    "**simga**: Sigma parameter in Gaussian function\n",
    "\n",
    "**R2**: R squared of how good is the Gaussian fit on the peak \n",
    "\n",
    "**isoshift**: Observed isotopic shift\n",
    "\n",
    "**theorshift**: Theoretical isotopic shift\n",
    "\n",
    "**masserror**: The difference between isoshift and theorshift\n",
    "\n",
    "**isoratio**: Division of I in isotopic peaks over I monoisotopic peak in the same scan number\n",
    "\n",
    "**rt**: Retention time in seconds\n",
    "\n",
    "**tic**: Total ion current\n",
    "\n",
    "**bpI**: Base peak intensity\n",
    "\n",
    "**bpMZ**: Base peak mz\n",
    "\n",
    "**ion**: 1-letter code of amino acid\n",
    "\n",
    "**n**: Number of the element in the ion (e.g. immonium ion of proline has 4 carbons) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For MS1 the monosiotopic peak is represented by the amino acid (proline, hydroxyproline or valine), which contains the most abundant versions of 1H, 16O, 14N, 12C. Because the isotopic variant are so rare we assume that the prevalence of 2H, 18H, 15N and 13C exist only in seperate amino acids. Therefore, we have one monoisotopic intensity and four different intensities for each isotope ratio (H<sub>1</sub>/H<sub>2</sub>, C<sub>12</sub>/C<sub>13</sub>, N<sub>14</sub>/N<sub>15</sub> and O<sub>16</sub>/O<sub>18</sub>).\n",
    "\n",
    "For MS2 we instead use the intensities of the fragment ions. The fragmentation processes is documented [here](http://www.matrixscience.com/help/fragmentation_help.html). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----- OLD STUFF BELOW -------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2020-11-10 Tuesday\n",
    "ToDo:\n",
    "- Check dinosaur, openms, skyline...\n",
    "- Skim (Cramer et al.)[https://www.researchgate.net/publication/47730346_Evaluation_of_Peak-Picking_Algorithms_for_Protein_Mass_Spectrometry]\n",
    "- Check how to vizualize and vizualise .mzML files to inspect.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2020-10-21 Webnesday\n",
    "### 14:27 More tools, pyopenms and previous run of conversion to .mzML was bad\n",
    "Previous run conversion run with msconvert was bad. Just reran the code without peakPicking filter.\n",
    "\n",
    "Checked out pyopenms. It works on my desktop, but there is an error related to QT5 on my laptop when running on spyder (why?). \n",
    "\n",
    "I got the suggestion to use dinosaur as a peakPicker and try different peakPicking filters from it. \n",
    "\n",
    "ToDo:\n",
    "- check if Percolator applicable at some step here?\n",
    "- check how to run dinosaur. \n",
    "- peakPick with dinosaur. \n",
    "- Script code to generate some exploratory plots with .mzML files.\n",
    "- Check if pandas-profiling can be useful. \n",
    "\n",
    "## 2020-10-18\n",
    "\n",
    "### 23:27 Finished data conversion to .mzML\n",
    "Just used msconvert to convert the .raw data to .mzML. The only filter used is pickPeaking msLevel=1-2. The end file size are much smaller than expected even though I've been told they are expected to be larger than the .raw files. Am I doing something wrong?\n",
    "\n",
    "I will try pyopenms to open these .mzML files.\n",
    "\n",
    "Setting up a Rmarkdown log for this project. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
